src/libAtomVM/bif.c

```diff
@@ -108,6 +108,45 @@ term bif_erlang_bit_size_1(Context *ctx, uint32_t fail_label, int live, term arg
     return term_from_int32(len);
 }
 
+bool get_sub_binary_slice(term bin_term, term pos_term, term len_term, struct SubBinarySlice *slice)
+{
+    int bin_size = term_binary_size(bin_term);
+    avm_int_t pos = term_to_int(pos_term);
+    avm_int_t len = term_to_int(len_term);
+
+    if (len < 0) {
+        pos += len;
+        len = -len;
+    }
+
+    if (UNLIKELY((pos < 0) || (pos > bin_size) || (pos + len > bin_size))) {
+        return false;
+    }
+
+    slice->pos = pos;
+    slice->len = len;
+    slice->heap_size = term_sub_binary_heap_size(bin_term, len);
+    return true;
+}
+
+term bif_erlang_binary_part_3(Context *ctx, uint32_t fail_label, int live, term arg1, term arg2, term arg3)
+{
+    VALIDATE_VALUE_BIF(fail_label, arg1, term_is_binary);
+    VALIDATE_VALUE_BIF(fail_label, arg2, term_is_integer);
+    VALIDATE_VALUE_BIF(fail_label, arg3, term_is_integer);
+
+    struct SubBinarySlice slice;
+    if (UNLIKELY(!get_sub_binary_slice(arg1, arg2, arg3, &slice))) {
+        RAISE_ERROR_BIF(fail_label, BADARG_ATOM);
+    }
+
+    if (UNLIKELY(memory_ensure_free_with_roots(ctx, slice.heap_size, live, ctx->x, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
```

Comment:
You need to add arg1 to ctx->[live], and pass live+1 with a comment saying you can do that because x has size MAX_REGS+1.

Reply:
I don't understand. If `live` is always zero, we'd clobber `x[1]` in that case and preserve nothing since roots length is still 0.

Reply:
Sorry for the confusion.

I meant:

```suggestion
    // We can always use ctx->x[live] as ctx->x has MAX_REGS+1 slots
    ctx->x[live] = arg1;
    if (UNLIKELY(memory_ensure_free_with_roots(ctx, slice.heap_size, live + 1, ctx->x, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
```

If `live` is 0, you can clobber x[1]. This is what `live` being 0 means.

Reply:
I don't understand the implementation then. In `maybe_call_native` we hardcode `live=0`. It's used in tail-calls (which is fine) and in `apply`. Why it's fine to clobber registers in `apply` case?

Reply:
I believe it's fine to clobber x regs in any NIF or external calls. My understanding is that x regs are used for args and results for NIF or external calls, and they do not need to be preserved. It is as if live was equal to 0. Caller does not even ensure that they are valid beyond argc. On return, only x[0] is considered valid.

Reply:
I agree with @pguyot, it works with both NIFs and BIFs.